tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/DefinedFunction.java

```diff
@@ -0,0 +1,212 @@
+/*
+  Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ ==============================================================================
+ */
+package org.tensorflow;
+
+import static org.tensorflow.internal.c_api.global.tensorflow.TF_FunctionName;
+import static org.tensorflow.internal.c_api.global.tensorflow.TF_FunctionSetAttrValueProto;
+import static org.tensorflow.internal.c_api.global.tensorflow.TF_GraphToFunction;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import org.bytedeco.javacpp.BytePointer;
+import org.bytedeco.javacpp.Pointer;
+import org.bytedeco.javacpp.PointerScope;
+import org.tensorflow.Graph.Reference;
+import org.tensorflow.internal.c_api.TF_Function;
+import org.tensorflow.internal.c_api.TF_Operation;
+import org.tensorflow.internal.c_api.TF_Output;
+import org.tensorflow.internal.c_api.TF_Status;
+import org.tensorflow.op.Ops;
+import org.tensorflow.proto.framework.AttrValue;
+
+/**
+ * A Function created from a TensorFlow {@link Graph}. Callable in Graph or Eager mode via the {@link
+ * Ops#callFunction(DefinedFunction, Operand[])} ops or the {@link #call(Ops, Operand[])} methods.
+ *
+ * <p>By default, resolves arguments and returns outputs by position.
+ * Names can be specified for the inputs and outputs (and used when calling) using {@link NamedGraphFunction} and {@link
+ * #create(Graph, String, boolean, Map, Map, String)}.
+ *
+ * Functions can also be created from {@link ConcreteFunction}s using {@link #create(ConcreteFunction, boolean)}.
+ */
+public class DefinedFunction implements AutoCloseable {
```

Comment:
I'm still not convinced that we need this extra class as part of the public API. Can't we have everything fronted by `ConcreteFunction`, eventually changing it to an interface that can have multiple implementations?

Reply:
I agree, but I'm running to access modifiers:
> When trying to make `DefinedFunction` package local, I ran into the issue where I'd like to have a `Ops.call(ConcreteFunction, Map<String, Operand<?>)` function, but the necessary methods (and `DefinedFunction`) won't be accessible there. There's a few options, as far as I can see:
> 
> 1. Only have a call method on `ConcreteFunction`, like `ConcreteFunction.call(Ops tf, Map<String, Operand<?>)`.
> 2. Have a `@Operator` Function class in the main package
> 3. Have a public call method on `ConcreteFunction` that takes `Scope` instead of `Ops`, call that from the operator class
> 4. Move `ConcreteFunction`, `DefinedFunction`, the Operator class, and any future function classes to a new package (`function`, presumably).
> 

4 is my preference, thoughts?

Reply:
So like we discussed, let's add endpoints to `ConcreteFunction` that accepts a `Scope` object and do the actual job. Anyway, most users will always go through the ops so it's ok to have these kind of endpoints in the public API of the core classes even if they are more meant for internal usage.

We also agreed that only `ConcreteFunction` should remain and probably be converted to an interface that has 2 different possible implementations: one with a graph (like the actual one) and one with a function graph (the new one). Maybe add these implementations in the `org.tensorflow.internal` package?

And last, can we check if it will be possible to build `ConcreteFunction` instances out of function graphs found in a saved model that has been loaded?

Reply:
I've started on this, and as far as I can tell, the only thing `ConcreteFunction` provides that the native functions doesn't is the ability to call with `Tensor`s, and even that's relocatable by using an temporary EagerSession.  So do we even need it?  The `Signature` builders can be ported just fine.  The only thing I could think of that would cause issues is Saving, which I haven't looked at enough to be able to tell.  If I understand it right it puts everything under a default named function anyways, so we could just create a graph and add the function to it.

Reply:
| So do we even need it? 

The goal is that we don't! Maybe I'll preserve `ConcreteFunction` as the name for the common function interface since it is still accurate without making breaking changes again. But if you think it is simple to do everything within a single class instead of one interface and two implementations, you can give it a try and I'll take a look at the PR but I have slightly a preference for the latter.

Reply:
Will do.  I'll start it without saving, and figure that out after.  I'm also not sure how variables are handled within `TF_Function`, which may cause some differences from `ConcreteFunction` (especially wrt saving), but I suspect that resource variables will be handled just fine, it's only the legacy ones that will have issues, so imo we don't need to worry about it.